### ISSUE TITLE
[ENH] Generate `__text_signature__` for cdef functions/classes

### ISSUE THREAD

**Issue by pitrou** [cython/cython#3873]
I have the following class declaration:
```cython
cdef class MinMaxOptions(FunctionOptions):
    def __init__(self, null_handling='skip'):
        # ...
```

Unfortunately, when compiled, the signature information for the constructor is lost:
```python
>>> pc.MinMaxOptions.__text_signature__
>>> pc.MinMaxOptions.__init__.__text_signature__
'($self, /, *args, **kwargs)'
```

**Comment by scoder** [comment]
This can be done with the `AutoDocTransform`, which we use to prepend signatures to docstrings. We could use the same string as `__text_signature__`.

Does this only apply to `__init__`, BTW?

**Comment by pitrou** [comment]
No, it seems to apply to all methods of `cdef` classes, though the symptoms are slightly different (`__init__` and `__new__` have a generic signature, other methods have no signature at all):
```python
>>> import pyarrow as pa
>>> pa.DataType.__init__.__text_signature__
'($self, /, *args, **kwargs)'
>>> pa.DataType.equals.__text_signature__
>>> pa.DataType.equals?
Docstring:
DataType.equals(self, other)
  # [ ... snip ... ]
>>> inspect.signature(pa.DataType.equals)
Traceback (most recent call last):
  # [ ... snip ... ]
ValueError: no signature found for builtin <method 'equals' of 'pyarrow.lib.DataType' objects>
```

**Comment by pitrou** [comment]
Note the goal for `__text_signature__` is to interpretable by `inspect.signature`. I hope Sphinx's autodoc will handle that.

**Comment by pitrou** [comment]
By the way, it seems `__text_signature__` is normally generated, on C functions, from the first line of the docstring:
https://github.com/python/cpython/blob/master/Modules/clinic/_bisectmodule.c.h#L6

**Comment by scoder** [comment]
The problem with special methods is that they lack a `__code__` object etc., and thus fail the test for `_signature_is_functionlike()` in `inspect.py`. They are not `CythonFunction` objects but the plain method wrappers that CPython generates for them.

If we could turn them into `CythonFunction`s, this issue would probably also go away. If we can't, then I'm not sure if we actually can change the `__text_signature__` attribute for them.

**Comment by scoder** [comment]
> By the way, it seems `__text_signature__` is normally generated, on C functions, from the first line of the docstring: https://github.com/python/cpython/blob/master/Modules/clinic/_bisectmodule.c.h#L6

This isn't referring to a special method, though. With the `binding` directive (which creates `CythonFunction`s and is enabled in Cython 3.0 and can be enabled in Cython 0.xy), `inspect.signature()` works for me.

**Comment by pitrou** [comment]
Are there any downsides to enabling the `binding` directive? The documentation doesn't seem to say anything.

**Comment by scoder** [comment]
It's possible that there are tiny differences that were only fixed in the 3.0 series and not in 0.29.x. But you shouldn't commonly run into those. In general, it should be ok to enable the directive.

**Comment by pitrou** [comment]
Well, I may be missing something, but I have the following Cython code (excerpt):
```cython
# cython: language_level = 3
# cython: binding = True

cdef class ModeOptions(FunctionOptions):
    __doc__ = f"""
    Options for the `mode` function.

    Parameters
    ----------
    n : int, default 1
        Number of distinct most-common values to return.
    {_skip_nulls_doc()}
    {_min_count_doc(default=0)}
    """

    def __init__(self, n=1, *, skip_nulls=True, min_count=0):
        self.wrapped.reset(new CModeOptions(n, skip_nulls, min_count))
```

Yet when compiled it doesn't seem to expose the required metadata:
```python console
>>> inspect.signature(_compute.ModeOptions)
Traceback (most recent call last):
  Input In [4] in <cell line: 1>
    inspect.signature(_compute.ModeOptions)
  File ~/miniconda3/envs/pyarrow/lib/python3.9/inspect.py:3113 in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File ~/miniconda3/envs/pyarrow/lib/python3.9/inspect.py:2862 in from_callable
    return _signature_from_callable(obj, sigcls=cls,
  File ~/miniconda3/envs/pyarrow/lib/python3.9/inspect.py:2397 in _signature_from_callable
    raise ValueError(
ValueError: no signature found for builtin type <class 'pyarrow._compute.ModeOptions'>

>>> inspect.signature(_compute.ModeOptions.__init__)
<Signature (self, /, *args, **kwargs)>
>>> inspect.signature(_compute.ModeOptions.__new__)
<Signature (*args, **kwargs)>
```

**Comment by scoder** [comment]
So, first of all (just to state the obvious), `ModeOptions` is an extension type. It is not itself affected by the `binding` directive, which replaces the standard CPython `PyCFunction` implementation of compiled Python functions with `cyfunction`, a C "subclass" of `PyCFunction` that adds Python compatibilty and introspection features.

Secondly, `__init__` is a special method, which CPython wraps as `PyCFunction` internally, and Cython does not change it or overwrite it after the fact. You can see that from the method's type. If your extension type had a non-special method, that would be introspectible.

As for `__new__`, that's the one in `object` that you see here.

**Comment by pitrou** [comment]
Ok, so `binding` would work with a regular method of an extension class, but not with `__init__`? I'm afraid that doesn't help me much in this case :-)

**Comment by Gouvernathor** [comment]
> This can be done with the AutoDocTransform, which we use to prepend signatures to docstrings. We could use the same string as __text_signature__.

This seems to be a sane solution. Is there a reason not to do that ?

**Comment by da-woods** [comment]
> > This can be done with the AutoDocTransform, which we use to prepend signatures to docstrings. We could use the same string as **text_signature**.
> 
> This seems to be a sane solution. Is there a reason not to do that ?

I can't even see a way of attaching `__doc__` to the `__init__` special method, so it's hard to see how to get this to work. (Could be missing something though)

**Comment by Gouvernathor** [comment]
Would it be easier to just attach it to the class ? That would be enough with what I understand of inspect.signature
